src/Illuminate/Cache/DatabaseStore.php

```diff
@@ -149,23 +145,32 @@ public function decrement($key, $value = 1)
      * @param  string  $key
      * @param  mixed  $value
      * @param  \Closure  $callback
-     * @return void
+     * @return int|bool
      */
     protected function incrementOrDecrement($key, $value, Closure $callback)
     {
-        $prefixed = $this->prefix.$key;
+        return $this->connection->transaction(function () use ($key, $value, $callback) {
+            $prefixed = $this->prefix.$key;
 
-        $cache = $this->table()->where('key', $prefixed)->lockForUpdate()->first();
+            $cache = $this->table()->where('key', $prefixed)->lockForUpdate()->first();
+
```

Comment:
But you still already read the old increment value before it was
incremented by another process.

On Wed, Dec 30, 2015 at 9:40 AM, Will Rowe <notifications@github.com> wrote:

> In src/Illuminate/Cache/DatabaseStore.php
> <https://github.com/laravel/framework/pull/11582#discussion_r48610497>:
>
> >
> > -        $cache = $this->table()->where('key', $prefixed)->lockForUpdate()->first();
> > +            $cache = $this->table()->where('key', $prefixed)->lockForUpdate()->first();
> > +
>
> I know that MySQL will prevent that from happening by using the updated
> value:
>
> SELECT ... LOCK IN SHARE MODE sets a shared mode lock on any rows that are
> read. Other sessions can read the rows, but cannot modify them until your
> transaction commits. If any of these rows were changed by another
> transaction that has not yet committed, your query waits until that
> transaction ends and then uses the latest values.
>
> You can reference that here
> <http://dev.mysql.com/doc/refman/5.5/en/innodb-locking-reads.html>. I'm
> not sure about the other database types though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/framework/pull/11582/files#r48610497>.
>